drivers/flash/soc_flash_nios2_qspi.c

```diff
@@ -0,0 +1,449 @@
+	u32_t word_to_write, padding, bytes_to_copy;
+
+	while (remaining_length > 0) {
+		/* initialize word to write to blank word */
+		word_to_write = 0xFFFFFFFF;
+
+		/* bytes to pad the next word that is written */
+		padding = 0;
+
+		/* number of bytes from source to copy */
+		bytes_to_copy = NIOS2_WRITE_BLOCK_SIZE;
+
+		/*
+		 * we need to make sure the write is word aligned
+		 * this should only be true at most 1 time
+		 */
+		if (0 != (write_offset & (NIOS2_WRITE_BLOCK_SIZE - 1))) {
+			/*
+			 * data is not word aligned calculate padding bytes
+			 * need to add before start of a data offset
+			 */
+			padding = write_offset & (NIOS2_WRITE_BLOCK_SIZE - 1);
+
+			/*
+			 * update variables to account
+			 * for padding being added
+			 */
+			bytes_to_copy -= padding;
+
+			if (bytes_to_copy > remaining_length) {
+				bytes_to_copy = remaining_length;
+			}
+
+			write_offset = write_offset - padding;
+
+			if (0 != (write_offset &
+					(NIOS2_WRITE_BLOCK_SIZE - 1))) {
+				return -EINVAL;
+			}
+		} else {
+			if (bytes_to_copy > remaining_length) {
+				bytes_to_copy = remaining_length;
+			}
+		}
+
+		/* prepare the word to be written */
+		memcpy(((void *)&word_to_write) + padding,
+				((void *)data) + buffer_offset, bytes_to_copy);
+
+		/* enable write */
+		IOWR_32DIRECT(qspi_dev->csr_base,
+				ALTERA_QSPI_CONTROLLER2_MEM_OP_REG,
+				MEM_OP_WRITE_EN_CMD);
+
+		/* write to flash 32 bits at a time */
+		IOWR_32DIRECT(qspi_dev->data_base, write_offset, word_to_write);
+
+		/* check whether write triggered a illegal write interrupt */
+		if ((IORD_ALTERA_QSPI_CONTROLLER2_ISR(qspi_dev->csr_base)
+			& ALTERA_QSPI_CONTROLLER2_ISR_ILLEGAL_WRITE_MASK) ==
+			ALTERA_QSPI_CONTROLLER2_ISR_ILLEGAL_WRITE_ACTIVE) {
+			/* clear register */
+			IOWR_ALTERA_QSPI_CONTROLLER2_ISR(qspi_dev->csr_base,
+				ALTERA_QSPI_CONTROLLER2_ISR_ILLEGAL_WRITE_MASK);
+			return -EIO; /* sector might be protected */
+		}
+
+		/* update offset and length variables */
+		buffer_offset += bytes_to_copy;
+		remaining_length -= bytes_to_copy;
+		write_offset = write_offset + NIOS2_WRITE_BLOCK_SIZE;
+	}
+
+	return 0;
+}
+
+static int flash_nios2_qspi_write(struct device *dev, off_t offset,
+				  const void *data, size_t len)
+{
+	struct flash_nios2_qspi_config *flash_cfg = dev->driver_data;
+	alt_qspi_controller2_dev *qspi_dev = &flash_cfg->qspi_dev;
+	u32_t block_offset, offset_in_block, length_to_write;
+	u32_t write_offset = offset; /* address of next byte to write */
+	u32_t buffer_offset = 0; /* offset into source buffer */
+	u32_t remaining_length = len; /* length of data left to be written */
+	int rc = 0, i;
+
+	k_sem_take(&flash_cfg->sem_lock, K_FOREVER);
+	/*
+	 * check if offset is word aligned and
+	 * length is with in the range
+	 */
+	if ((data == NULL) || ((offset + len) > qspi_dev->data_end) ||
+			(0 != (write_offset &
+			       (NIOS2_WRITE_BLOCK_SIZE - 1)))) {
+		printk("nios2-qspi: read failed at offset %ld\n", offset);
```

Comment:
consider using Zephyr's error logging macros instead of printk() directly

Reply:
Ok.